02-peer-protocol.md

```diff
@@ -49,11 +68,468 @@ transaction is confirmed are also not persistent - until you know the script
+        +-------+                       +-------+
+
+#### *initiator* and *non-initiator*
+
+A the *initiator* contributes 2 inputs and an output that they
+then remove.  B, the *non-initiator*, contributes 1 input and an output,
+but waits until A adds a second input before contributing.
+
+Note that if A does not send a second input, the negotiation will end without
+B's contributions.
+
+        +-------+                         +-------+
+        |       |--(1)- tx_add_input ---->|       |
+        |       |<-(2)- tx_complete ------|       |
+        |       |--(3)- tx_add_output --->|       |
+        |       |<-(4)- tx_complete ------|       |
+        |       |--(5)- tx_add_input ---->|       |
+        |   A   |<-(6)- tx_add_input -----|   B   |
+        |       |--(7)- tx_remove_output >|       |
+        |       |<-(8)- tx_add_output ----|       |
+        |       |--(9)- tx_complete ----->|       |
+        |       |<-(10) tx_complete ------|       |
+        +-------+                         +-------+
+
+### The `tx_add_input` Message
+
+This message contains a transaction input.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u32`:`sequence`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input currently added to the
+    transaction
+  - MUST NOT re-transmit inputs it has received from the peer
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received inputs to the transaction
+  - MUST fail the negotiation if:
+    - the `prevtx` and `prevtx_vout` are identical to a previously added
+      (and not removed) input's
+    - `prevtx` is not a valid transaction
+    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
+    - the `scriptPubKey` of the `prevtx_vout` output of `prevtx` is not exactly a 1-byte push opcode (for the numeric values `0` to `16`) followed by a data push between 2 and 40 bytes
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - if has received 4096 `tx_add_input` messages during this negotiation
+
+#### Rationale
+
+Each node must know the set of the transaction inputs. The *non-initiator*
+MAY omit this message.
+
+`serial_id` is a randomly chosen number which uniquely identifies this input.
+Inputs in the constructed transaction are sorted by `serial_id`.
+
+`prevtx` is the serialized transaction that contains the output
+this input spends. Used to verify that the input is non-malleable.
+
+`prevtx_vout` is the index of the output being spent.
+
+`sequence` is the sequence number of this input.  Must be less than
+4294967294 (0xFFFFFFFE). See BIP125.
```

Comment:
a transaction signals rbf if **at least 1 input** has `nSequence < 0xfffffffe`
only forcing one input's nSequence to be below 0xfffffffe feels more correct

also, a link to bip125 could be added here:
`https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#summary`

Reply:
I'm not sure we should mention BIP 125, as bitcoin core is in the process of removing references to it since what's actually implemented today doesn't match BIP 125...